src/sourcegraph.d.ts

```diff
@@ -338,42 +337,251 @@ declare module 'sourcegraph' {
+     */
+    export type ViewComponent = CodeEditor
+
+    /**
+     * A text document decoration changes the appearance of a range in the document and/or adds other content to
+     * it.
+     */
+    export interface TextDocumentDecoration {
+        /** The range that the decoration applies to. */
+        range: Range
+
+        /**
+         * If true, the decoration applies to all lines in the range (inclusive), even if not all characters on the
+         * line are included.
+         */
+        isWholeLine?: boolean
+
+        /** Content to display after the range. */
+        after?: DecorationAttachmentRenderOptions
+
+        /** The CSS background-color property value for the line. */
+        backgroundColor?: string
+
+        /** The CSS border property value for the line. */
+        border?: string
+
+        /** The CSS border-color property value for the line. */
+        borderColor?: string
+
+        /** The CSS border-width property value for the line. */
+        borderWidth?: string
+    }
+
+    /** A decoration attachment adds content after a {@link TextDocumentDecoration}. */
+    export interface DecorationAttachmentRenderOptions {
+        /** The CSS background-color property value for the attachment. */
+        backgroundColor?: string
+
+        /** The CSS color property value for the attachment. */
+        color?: string
+
+        /** Text to display in the attachment. */
+        contentText?: string
+
+        /** Tooltip text to display when hovering over the attachment. */
+        hoverMessage?: string
+
+        /** If set, the attachment becomes a link with this destination URL. */
+        linkURL?: string
+    }
+
+    /**
+     * A text editor for code files (as opposed to a rich text editor for documents or other kinds of file format
+     * editors).
+     */
+    export interface CodeEditor {
+        /** The type tag for this kind of {@link ViewComponent}. */
+        type: 'CodeEditor'
+
+        /**
+         * The text document that is open in this editor.
+         */
+        readonly document: TextDocument
+
+        /**
+         * Draw decorations on this editor.
+         *
+         * @todo Implement a "decoration type" as in VS Code to make deltas more efficient.
+         * @param decorationType Currently unused. Always pass `null`.
+         */
+        setDecorations(decorationType: null, decorations: TextDocumentDecoration[]): void
+    }
+
+    /**
+     * The client application that is running the extension.
+     */
+    export namespace app {
+        /**
+         * The currently active window, or `undefined`. The active window is the window that has focus, or when
+         * none has focus, the window that was most recently focused.
+         */
+        export const activeWindow: Window | undefined
+
+        /**
+         * All application windows that are accessible by the extension.
+         *
+         * @readonly
```

Comment:
Instead of marking things `readonly` in comments, why not make it `readonly` in type information? I.e. `ReadonlyArray<Window>`.

Reply:
I assign EXTREME value to making this file human-readable. I believe that `Window[]` is much more readable than `ReadonlyArray<Window>`. The latter will make people think this API is complex. Adding a `@readonly` has low cost, though.

I understand that there is a counterargument of "it is better to communicate that it is a readonly array", but I think way fewer people will be confused by that than if all arrays were `ReadonlyArray<...>`.